src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -126,96 +126,154 @@ public Grid (int zoom, int width, int height, int north, int west) {
         this.grid = new double[width][height];
     }
 
+    public static class PixelWeight {
+        int x;
+        int y;
+        double weight;
+
+        private PixelWeight (int x, int y, double weight){
+            this.x = x;
+            this.y = y;
+            this.weight = weight;
+        }
+
+        public int getX() {
+            return x;
+        }
+        public int getY() {
+            return y;
+        }
+        public double getWeight() {
+            return weight;
+        }
+    }
+
+    /**
+     * A functional interface (callback interface) that processes one pixel weight at a time.
+     * This avoids storing massive lists of intermediate results that have been known to eat memory on the server.
+     * This happens when processing geometries that contain a huge amount of pixels, which a user might include in
+     * a shapefile.
+     */
+    public interface PixelWeightCallback {
+        void handlePixelWeight (int x, int y, double weight);
+    }
+
+    /**
+     * Stream pixel weights using a callback, rather than keeping them in memory.
+     */
+    public void streamPixelWeights (Geometry geometry, boolean relativeToPixels, PixelWeightCallback pixelWeightCallback) {
+        getPixelWeights(geometry, relativeToPixels, pixelWeightCallback);
+    }
+
     /**
      * Version of getPixelWeights which returns the weights as relative to the total area of the input geometry (i.e.
      * the weight at a pixel is the proportion of the input geometry that falls within that pixel.
      */
-    public TObjectDoubleMap<int[]> getPixelWeights (Geometry geometry) {
-        return getPixelWeights(geometry, false);
+    public ArrayList<PixelWeight> getPixelWeights (Geometry geometry) {
```

Comment:
Field and return types should usually be an interface type (List) rather than the implementation type (ArrayList).